TESTS/mbed_drivers/ticker/main.cpp

```diff
@@ -147,14 +203,167 @@ void test_case_2x_callbacks() {
 
 }
 
+/** Test many tickers run simultaneously
+
+    Given an 16 tickers
+    When schedule them
+    Then tickers properly execute callbacks
+ */
+void test_multi_ticker(void)
+{
+    TestTicker ticker[TICKER_COUNT];
+
+    multi_counter = 0;
+    for(int i = 0; i < TICKER_COUNT; i++) {
+        ticker[i].attach_us(callback(wait_callback_multi), 100 * 1000);
+    }
+
+    Thread::wait(105);
+    for(int i = 0; i < TICKER_COUNT; i++) {
+            ticker[i].detach();
+    }
+
+    TEST_ASSERT_EQUAL(TICKER_COUNT, multi_counter);
+
+    multi_counter = 0;
+    for(int i = 0; i < TICKER_COUNT; i++) {
+        ticker[i].attach_us(callback(wait_callback_multi), (100 + i) * 1000);
+    }
+
+    Thread::wait(120);
+    for(int i = 0; i < TICKER_COUNT; i++) {
+        ticker[i].detach();
+    }
+    TEST_ASSERT_EQUAL(TICKER_COUNT, multi_counter);
+}
+
+/** Test multi callback time
+
+    Given an Ticker
+    When schedule callback
+    Then ticker properly execute callback multiple time
+ */
+void test_multi_call_time(void)
+{
+    TestTicker ticker;
+    int32_t ret;
+    Timer timer;
+    int time_diff;
+
+    for(int i = 0; i < 10; i++) {
+        ticker_callback_flag = false;
+        timer.reset();
+
+        timer.start();
+        ticker.attach_us(callback(wait_callback_sem), 100000);
+        ret = ticker_callback_sem.wait();
+        timer.stop();
+        time_diff = timer.read_us();
+
+        TEST_ASSERT_EQUAL(true, ticker_callback_flag);
+        TEST_ASSERT_TRUE(ret > 0);
+        TEST_ASSERT_UINT32_WITHIN(TOLERANCE, 100000, time_diff);
+    }
+}
+
+/** Test if detach cancel scheduled callback event
+
+    Given an Ticker with scheduled callback
+    When schedule is cancelled
+    Then callback should not be called anymore
+ */
+void test_detach(void)
+{
+    TestTicker ticker;
+    int32_t ret;
+    ticker_callback_flag = false;
+    ticker.attach(callback(wait_callback_sem), 0.1f);
+
+    ret = ticker_callback_sem.wait();
+    TEST_ASSERT_EQUAL(true, ticker_callback_flag);
+    TEST_ASSERT_TRUE(ret > 0);
+    ticker_callback_flag = false;
+
+    ticker_callback_sem.wait();
+    ticker.detach(); /* cancel */
+    TEST_ASSERT_EQUAL(true, ticker_callback_flag);
+    TEST_ASSERT_TRUE(ret > 0);
+    ticker_callback_flag = false;
+
+    ret = ticker_callback_sem.wait(300);
+    TEST_ASSERT_EQUAL(false, ticker_callback_flag);
+    TEST_ASSERT_EQUAL(0, ret);
+}
+
+/** Test single callback time via attach
+
+    Given an Ticker
+    When schedule callback
+    Then ticker properly execute callback
+ */
+template<us_timestamp_t DELAY_US>
+void test_attach_time(void)
+{
+    TestTicker ticker;
+    int32_t ret;
+    Timer timer;
+    ticker_callback_flag = false;
+
+    timer.start();
+    ticker.attach(callback(wait_callback_sem), ((float)DELAY_US) / 1000000.0f);
+    ret = ticker_callback_sem.wait();
+    timer.stop();
+    ticker.detach();
+
+    int time_diff = timer.read_us();
+    printf("SystemCoreClock: %d\n", SystemCoreClock);
```

Comment:
do we want all these print by default (not for instance only for debug builds?)

Reply:
fixed